src/docfx/lib/YamlUtility.cs

```diff
@@ -0,0 +1,162 @@
+        /// </summary>
+        public static T Deserialize<T>(string input)
+        {
+            return Deserialize<T>(new StringReader(input));
+        }
+
+        /// <summary>
+        /// Deserialize From TextReader
+        /// </summary>
+        public static T Deserialize<T>(TextReader reader)
+        {
+            var json = Deserialize(reader);
+            return json.ToObject<T>(s_jsonSerializer);
+        }
+
+        /// <summary>
+        /// Deserialize to JToken From string
+        /// </summary>
+        public static JToken Deserialize(string input)
+        {
+            return Deserialize(new StringReader(input));
+        }
+
+        /// <summary>
+        /// Deserialize to JToken from TextReader
+        /// </summary>
+        public static JToken Deserialize(TextReader reader)
+        {
+            var stream = new YamlStream();
+
+            stream.Load(reader);
+            if (stream.Documents.Count != 1)
+            {
+                throw new NotSupportedException("Does not support mutiple YAML documents");
+            }
+            return ToJson(stream.Documents[0].RootNode);
+
+            JToken ToJson(YamlNode node)
+            {
+                if (node is YamlScalarNode scalar)
+                {
+                    if (scalar.Style == ScalarStyle.Plain)
+                    {
+                        if (string.IsNullOrWhiteSpace(scalar.Value))
+                        {
+                            return null;
+                        }
+                        if (long.TryParse(scalar.Value, out var n))
+                        {
+                            return new JValue(n);
+                        }
+                        if (double.TryParse(scalar.Value, out var d))
+                        {
+                            return new JValue(d);
+                        }
+                        if (bool.TryParse(scalar.Value, out var b))
+                        {
+                            return new JValue(b);
+                        }
+                    }
+                    return new JValue(scalar.Value);
+                }
+                if (node is YamlMappingNode map)
+                {
+                    var obj = new JObject();
+                    foreach (var (key, value) in map)
+                    {
+                        obj[key.ToString()] = ToJson(value);
+                    }
+                    return obj;
+                }
+                if (node is YamlSequenceNode seq)
+                {
+                    var arr = new JArray();
+                    foreach (var item in seq)
+                    {
+                        arr.Add(ToJson(item));
+                    }
+                    return arr;
+                }
+                throw new NotSupportedException($"Unknown yaml node type {node.GetType()}");
+            }
+        }
+
+        class JsonContractResolver : DefaultContractResolver
```

Comment:
This ContractResolver was to support deserializing into readonly fields and properties in the form of:

- public readonly string Field
- public string GetOnlyProperty { get; }

Since our style does not encourage *readonly fields*, lets drop *readonly fields* and add the following property tests:

1. Can deserialize into get only property
2. When deserialize into get only property with default value, `null` and `undefined` does not override default value. E.g. when loading `public string A { get; } = "default"` from `{ "a": null }` or `{ }`, the result object should be `{ "a": "default" }`

Reply:
share the same logic with JsonUtility that both support readonly properties or both not...

---
In reply to: [180948737](https://github.com/dotnet/docfx/pull/2626#discussion_r180948737) [](ancestors = 180948737)

Reply:
there is not a way to deserialize a get-only property, because there is not a setter for this property, value cannot be set to this property even if we set `Writable` to true.

---
In reply to: [180956068](https://github.com/dotnet/docfx/pull/2626#discussion_r180956068) [](ancestors = 180956068,180948737)